addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js

```diff
@@ -515,6 +835,131 @@ const ColorPaletteWidget = Widget.extend({
         ev.stopPropagation();
         this._selectTabFromButton(ev.currentTarget);
     },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onGradientSliderClick(ev) {
+        ev.stopPropagation();
+        this._activateGradientSlider($(ev.target));
+        this._updateGradient();
+    },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onGradientPreviewClick(ev) {
+        ev.stopPropagation();
+        const offset = ev.offsetX;
+        const width = parseInt(window.getComputedStyle(ev.target).width, 10);
+        const position = 100 * offset / width;
+
+        let previousColor;
+        let nextColor;
+        for (const slider of $(this.gradientEditorParts.sliders).find('input')) {
+            if (slider.value < position) {
+                previousColor = $(slider).data('color');
+            } else {
+                nextColor = $(slider).data('color');
+                break;
+            }
+        }
+        let color;
+        if (previousColor && nextColor) {
+            previousColor = ColorpickerWidget.convertCSSColorToRgba(previousColor);
+            nextColor = ColorpickerWidget.convertCSSColorToRgba(nextColor);
+            const red = Math.round((previousColor.red + nextColor.red) / 2);
+            const green = Math.round((previousColor.green + nextColor.green) / 2);
+            const blue = Math.round((previousColor.blue + nextColor.blue) / 2);
+            const opacity = Math.round((previousColor.opacity + nextColor.opacity) / 2);
+            color = `rgba(${red}, ${green}, ${blue}, ${opacity / 100})`;
+        } else {
+            color = nextColor || previousColor || 'rgba(128, 128, 128, 0.5)';
+        }
+
+        const $slider = this._createGradientSlider(position, color);
+        this._activateGradientSlider($slider);
+        this._updateGradient();
+    },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onPanelClick(ev) {
+        // Ignore to avoid closing popup.
+        ev.stopPropagation();
+    },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onGradientInputBlur(ev) {
+        this._updateGradient();
+    },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onGradientButtonClick(ev) {
+        const $buttons = $(ev.target).closest('span').find('we-button');
+        $buttons.removeClass('active');
+        $(ev.target).closest('we-button').addClass('active');
+        this._updateGradient();
+    },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onGradientButtonEnter(ev) {
+        ev.stopPropagation();
+        const $activeButton = $(ev.target).closest('span').find('we-button.active');
+        const $buttons = $(ev.target).closest('span').find('we-button');
+        $buttons.removeClass('active');
+        $(ev.target).closest('we-button').addClass('active');
+        this._updateGradient(true);
+        $buttons.removeClass('active');
+        $activeButton.addClass('active');
```

Comment:
General comment: I am afraid with so much code dedicated to activatate and/or hide stuff in event handler. That makes me think this has been coded in a non-descriptive UI way. For me, the UI should only enable / show / hide stuff according to the gradient that is being applied on the edited DOM, I mean:

- The users previews stuff -> the UI has nothing to update
- The user applies stuff (by releasing its mouse or clicking a button or whatever) -> This is "saved" on the DOM -> The related option calls its `updateUI` -> `_computeWidgetState` -> `widget.setValue` flow, ending with the colorpalette receiving "Hey the gradient XXX is applied" => Ok, I know what I must hide and update.

This ensures that:
- No code is duplicated -> No "Oh you clicked that button? I will set the active class on it, not caring what it did" -> No: "You clicked that button? Ok I'll apply what is does to the DOM, then the DOM will answer me how I must update my UI accordingly **like I would have updated my UI when first clicking on a DOM with a gradient already applied on it**"
- The UI is synchronized with what the DOM actually shows **by force**, as the UI only shows what it understand the DOM looks like and not "what it think the user wanted to do by clicking stuff"

Maybe I am missing something but I don't think so, here is a symptom of the problem:
- Click on a snippet
- Go to gradients -> the "Custom" button is gray -> Click on it
- Add a gradient
- Open the colorpicker again -> Remove the gradient with the trash icon
- Open the colorpicker again -> The "Custom" button is still showing a gradient -> The UI is not synchronized with what the DOM actually looks like.

Reply:
The missing "reset" of the "Custom" button was actually just a missing update of the button background in case the DOM said there was no gradient.

The problem this method has is that computing a new gradient string is done from the state of the options. When hovering the options button, to obtain such values I temporarily pretend the hovered button is activated only to compute the new gradient string from the options. I could try to come up with a way to compute the gradient string taking into account an override of some options, but it feels like it will make the whole thing even more complex.

But indeed there is also an incorrect impact on the options from applying a new gradient instead of doing it from the DOM's feedback - working on this now. Thank you for the great explanation.

Reply:
To get feedback from the DOM, I added calls to `setSelectedColor` from outside:
- after `_processAndApplyColor` for toolbar-related use
- when the `colorPreviewEl` get updated
At least it happens outside of the palette.

Reply:
`_processAndApplyColor` uses `execCommand('applyColor', ...)` which already kept track of history steps => do not additionally touch history on color pick, and suspend history updates during previews.

Also restoring lost selection before calling `_processAndApplyColor`.